src/passes/CodeFolding.cpp

```diff
@@ -249,8 +249,12 @@ struct CodeFolding : public WalkerPass<ControlFlowWalker<CodeFolding>> {
       // remove if (4 bytes), remove one arm, add drop (1), add block (3),
       // so this must be a net savings
       markAsModified(curr);
+      auto* ifTrue = curr->ifTrue;
+      if (curr->type == Type::unreachable && curr->ifTrue->type.isConcrete()) {
+        ifTrue = builder.makeDrop(ifTrue);
+      }
       auto* ret =
-        builder.makeSequence(builder.makeDrop(curr->condition), curr->ifTrue);
+        builder.makeSequence(builder.makeDrop(curr->condition), ifTrue);
```

Comment:
Reading this code and the updated test following it, I still don't understand the change. The code looks valid before and after, and perhaps it is slightly more optimal to add a drop here rather than to wait for another pass to do it, but is it worth the complexity? Is there another reason I haven't considered?

Reply:
We get a validation error without inserting the drop. Without this change, we parse this wat:

```
(func $foo (result i32)
  i32.const 0
  unreachable
  if (result i32)
    i32.const 1
  else
    i32.const 2
  end
)
```
into this IR:
```
(func $foo (result i32)
  (i32.const 0) ;; Validation error!
  (if
    (unreachable)
    (then
       (i32.const 1)
    )
    (else 
       (i32.const 2)
    )
  )
)
```
This IR is invalid because the function body block has a non-final item with a concrete type. The IR should have contained a drop of that `i32.const 0`.

Reply:
Oh, this response wasn't specific to the CodeFolding logic, but it has a similar problem.

Down below we have `ref->finalize(curr->type)`. If `curr->type` is `Type::unreachable`, then it was previously the case that `curr->ifTrue` must have been type `none` or `unreachable`, but now it can be any type. Finalizing a sequence block where the second item has type e.g. `i32` as `unreachable` is not valid, so to fix it we have to drop that second item.